### PR TITLE
Tokens cleanup

### PR DESCRIPTION
**Pull request by jeremyhuiskamp**
Refactoring based on code review, and building in expiry support.

NB: you can no longer opt out of expiry except by choosing a date ridiculously far in the future.

### CODE REVIEW

File: tokens/tokens_test.go [hunk 1]
```diff
@@ -149,8 +153,11 @@ func replaceRandomRune(in string) string {
 	if payloadEnd == -1 {
 		panic("incorrect token")
 	}
-	// we need a random position in the payload
-	n := mrand.Intn(payloadEnd-payloadStart) + payloadStart + 1
+	// We need a random position in the payload.  -10 because
+	// modifications near the end often *don't* cause a problem,
+	// presumably because they are meaningless padding.  We're
+	// assuming all payloads will be at least 10 characters long.
+	n := mrand.Intn(payloadEnd-payloadStart-10) + payloadStart + 1
```
Comment (jeremyhuiskamp) [inline tokens/tokens_test.go]: This concerns me a bit, but I am assuming that the `jose` implementation is correct and that there's an acceptable explanation for what I've been seeing.  If anyone has concerns, we can take some more time to look into it.

File: tokens/tokens.go [hunk 2]
```diff
@@ -100,9 +108,27 @@ func (v *Verifier) Parse(token string, payload interface{}) (err error) {
...
+
+// Decode validates the token at the current time and extracts the payload from it.
+//
+// The payload must be a pointer to the interface and must be JSON decodable.
+func (v *Decoder) Decode(token string, payload interface{}) error {
+	return v.DecodeAsOf(time.Now(), token, payload)
+}
+
+type tokenContent struct {
+	Expiry  time.Time   `json:"expiry"`
```
Comment (alpe) [inline tokens/tokens.go]: If we stick with the rfc this is `exp`
https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32#section-4.1.4

Comment (jeremyhuiskamp) [inline tokens/tokens.go]: We're not compliant with that definition of `exp` (we serialize as an ISO8601 date while `exp` apparently requires a number).  As we have not made any effort to implement the claims spec, and as "expiry" is a better name than "exp" (in my opinion anyway), I'm inclined to not bother.

**Comment by jeremyhuiskamp** [comment]
Re validating against a different source, you make a good point.  I don't want to hold up moving forward with this, but later I will try to generate a token with a different library and incorporate it into a unit test.  Of course the content of the token is not standard, but I can hand-create a json body.

For now though, being able to validate our own tokens is sufficient.  If we have to break compatibility later, the only harm we'll be doing is dropping short-lived tokens of recipients who happen to be checking out at the point of upgrade (yes, that sucks, but I'm sure we do it often enough already).

**Pull request merged**